### PR TITLE
add credits plugin

### PR DESCRIPTION
**Pull request by mpwsh**

### CODE REVIEW

File: charts/hub-gateway/plugins/credits.lua [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    name = "credits",
+    schema = schema
+}
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.access(conf, ctx)
+    local org_cookie_name = conf.org_cookie_name or "_hub_org"
```
Comment (kespinola) [inline charts/hub-gateway/plugins/credits.lua:71]: This is good for browser based request but also need to apply to API where the client application has the organization.

File: charts/hub-gateway/plugins/credits.lua [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    local data, err = json.decode(res.body)
+    if not data then
+        return 500, json.encode({ message = err })
+    end
+
+    if not data.balance then
+        return 500, json.encode({ message = err })
+    end
+
+    if conf.expose_user_id then
```
Comment (kespinola) [inline charts/hub-gateway/plugins/credits.lua:112]:
```suggestion
    if conf.expose_credit_balance then
```

File: charts/hub-gateway/plugins/credits.lua [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        core.request.set_header(ctx, "X-CREDIT-BALANCE", data.balance)
+        core.response.set_header(ctx, "X-CREDIT-BALANCE", data.balance)
```
Comment (kespinola) [inline charts/hub-gateway/plugins/credits.lua:114]: X-CREDIT-BALANCED repeated twice.

Comment (kespinola) [reply charts/hub-gateway/plugins/credits.lua:114]: We also need the X-Organization-ID. To compose the plugins may be a good candidate for a dedicated plugin to consilated the cookie or client app lookup logic.

**Review by kespinola — CHANGES_REQUESTED** [review]
On a good track. Left some notes.

File: charts/hub-gateway/plugins/credits.lua (hunk 3 above)
Comment (mpwsh) [reply charts/hub-gateway/plugins/credits.lua:114]: > X-CREDIT-BALANCED repeated twice.

Its `core.request` and `core.response`, so we also reply back the credit balance to the user through headers.

I think its cool to let them know their balance constantly if they are not using Hub through the UI.


> We also need the X-Organization-ID. To compose the plugins may be a good candidate for a dedicated plugin to consilated the cookie or client app lookup logic.

Yea, agree. I'll add a `organization` plugin in the PR to move `x-client-owner-id` and `_hub_org` cookie to `x-organization-id` header

File: charts/hub-gateway/plugins/credits.lua [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    if not operation or operation == "query" then
+        return
+    end
+
+    local input = json.decode(core.request.get_body())
+    local graphql_query = input['query']
+
+    -- Check if query is createOrganization
+    if graphql_query:match("createOrganization") then
+        core.log.error("Creating Organization. Unable to continue.")
```
Comment (kespinola) [inline charts/hub-gateway/plugins/credits.lua:80]: Don't think this should be an error log. More like info. And message "Skipping credits plugin - Creating Organization"

**Review by kespinola — APPROVED** [review]
really nice.

**Pull request merged**